File: Helpers/src/main/java/com/dg/helpers/JSONHelper.java
package com.dg.helpers;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.lang.reflect.Array;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

/**
 * Created by Daniel Cohen Gindi (danielgindi@gmail.com)
 */
public class JSONHelper
{
    public static String[] toStringArray(JSONArray json)
    {
        if (json == null) return null;

        int length = json.length();
        String[] array = new String[length];

        for (int i = 0; i < length; i++)
        {
            Object value = json.opt(i);
            array[i] = value == null ? null : value instanceof String ? (String)value : String.valueOf(value);
        }

        return array;
    }

    public static ArrayList toArray(JSONArray json)
    {
        if (json == null) return null;

        int length = json.length();
        ArrayList<Object> array = new ArrayList<>(length);

        for (int i = 0; i < length; i++)
        {
            Object value = json.opt(i);
            if (value instanceof JSONArray)
            {
                array.add(toArray((JSONArray)value));
            }
            else if (value == JSONObject.NULL)
            {
                array.add(null);
            }
            else if (value instanceof JSONObject)
            {
                array.add(toMap((JSONObject) value));
            }
            else
            {
                array.add(value);
            }
        }

        return array;
    }

    public static <T> T[] toArray(JSONArray json, Class<T> typeClass)
    {
        if (json == null) return null;

        int length = json.length();
        ArrayList<Object> array = new ArrayList<>(length);

        for (int i = 0; i < length; i++)
        {
            Object value = json.opt(i);
            if (value instanceof JSONArray)
            {
                array.add(toArray((JSONArray)value));
            }
            else if (value == JSONObject.NULL)
            {
                array.add(null);
            }
            else if (value instanceof JSONObject)
            {
                array.add(toMap((JSONObject) value));
            }
            else
            {
                array.add(value);
            }
        }

        @SuppressWarnings("unchecked")
        T[] result = (T[]) Array.newInstance(typeClass, length);
        return array.toArray(result);
    }

    public static Map<String, Object> toMap(JSONObject json)
    {
        if (json == null) return null;

        Map<String, Object> map = new HashMap<>(json.length());

        Iterator<String> keys = json.keys();
        while (keys.hasNext())
        {
            String name = keys.next();

            Object value = json.opt(name);
            if (value instanceof JSONArray)
            {
                map.put(name, toArray((JSONArray) value));
            }
            else if (value == JSONObject.NULL)
            {
                map.put(name, null);
            }
            else if (value instanceof JSONObject)
            {
                map.put(name, toMap((JSONObject) value));
            }
            else
            {
                map.put(name, value);
            }
        }

        return map;
    }

    /**
     * Wraps the given object if necessary.
     *
     * <p>If the object is null, returns {@code JSONObject.NULL}.
     * If the object is a {@code JSONArray} or {@code JSONObject}, no wrapping is necessary.
     * If the object is {@code JSONObject.NULL}, no wrapping is necessary.
     * If the object is an array or {@code Collection}, returns an equivalent {@code JSONArray}.
     * If the object is a {@code Map}, returns an equivalent {@code JSONObject}.
     * If the object is a primitive wrapper type or {@code String}, returns the object.
     * Otherwise if the object is from a {@code java} package, returns the result of {@code toString}.
     * If wrapping fails, returns null.
     */
    public static Object wrap(Object o)
    {
        if (o == null)
        {
            return JSONObject.NULL;
        }

        if (o instanceof JSONArray || o instanceof JSONObject)
        {
            return o;
        }

        if (o.equals(JSONObject.NULL))
        {
            return o;
        }

        try
        {
            if (o instanceof Collection)
            {
                JSONArray jsonArray = new JSONArray();
                for (Object entry : (Collection)o)
                {
                    jsonArray.put(wrap(entry));
                }
                return jsonArray;
            }

            if (o.getClass().isArray())
            {
                final int length = Array.getLength(o);
                JSONArray jsonArray = new JSONArray();
                for (int i = 0; i < length; ++i)
                {
                    jsonArray.put(wrap(Array.get(o, i)));
                }

                return jsonArray;
            }

            if (o instanceof Map)
            {
                JSONObject object = new JSONObject();

                for (Map.Entry<?, ?> entry : ((Map<?, ?>)o).entrySet())
                {
                    String key = (String) entry.getKey();

                    if (key == null)
                    {
                        continue;
                    }

                    try
                    {
                        object.put(key, wrap(entry.getValue()));
                    }
                    catch (JSONException e)
                    {
                        e.printStackTrace();
                    }
                }

                return object;
            }

            if (o instanceof Boolean ||
                    o instanceof Byte ||
                    o instanceof Character ||
                    o instanceof Double ||
                    o instanceof Float ||
                    o instanceof Integer ||
                    o instanceof Long ||
                    o instanceof Short ||
                    o instanceof String) {
                return o;
            }

            if (o.getClass().getPackage().getName().startsWith("java."))
            {
                return o.toString();
            }

        }
        catch (Exception ignored)
        {
        }
        return null;
    }
}

File: Helpers/src/main/java/com/dg/helpers/JSONHelper.kt
package com.dg.helpers

import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

import java.util.ArrayList
import java.util.HashMap

@Suppress("unused")
object JSONHelper
{
    fun toStringArray(json: JSONArray?): Array<String?>?
    {
        if (json == null) return null

        val length = json.length()
        val array = arrayOfNulls<String>(length)

        for (i in 0 until length)
        {
            val value = json.opt(i)
            array[i] = if (value == null) null else value as? String ?: value.toString()
        }

        return array
    }

    fun toArray(json: JSONArray?): ArrayList<*>?
    {
        if (json == null) return null

        val length = json.length()
        val array = ArrayList<Any?>(length)

        for (i in 0 until length)
        {
            val value = json.opt(i)
            when
            {
                value is JSONArray -> array.add(toArray(value))
                value === JSONObject.NULL -> array.add(null)
                value is JSONObject -> array.add(toMap(value))
                else -> array.add(value)
            }
        }

        return array
    }

    inline fun <reified T> toArray(list: List<*>): Array<T>
    {
        @Suppress("UNCHECKED_CAST")
        return (list as List<T>).toTypedArray()
    }

    inline fun <reified T> toArray(json: JSONArray?): Array<T>?
    {
        if (json == null) return null

        val length = json.length()
        val array = ArrayList<Any?>(length)

        for (i in 0 until length)
        {
            val value = json.opt(i)
            when
            {
                value is JSONArray -> array.add(toArray(value))
                value === JSONObject.NULL -> array.add(null)
                value is JSONObject -> array.add(toMap(value))
                else -> array.add(value)
            }
        }

        return toArray<T>(array)
    }

    fun toMap(json: JSONObject?): Map<String, Any?>?
    {
        if (json == null) return null

        val map = HashMap<String, Any?>(json.length())

        val keys = json.keys()
        while (keys.hasNext())
        {
            val name = keys.next()

            val value = json.opt(name)
            when
            {
                value is JSONArray -> map[name] = toArray(value)
                value === JSONObject.NULL -> map[name] = null
                value is JSONObject -> map[name] = toMap(value)
                else -> map[name] = value
            }
        }

        return map
    }

    /**
     * Wraps the given object if necessary.
     *
     *
     * If the object is null, returns `JSONObject.NULL`.
     * If the object is a `JSONArray` or `JSONObject`, no wrapping is necessary.
     * If the object is `JSONObject.NULL`, no wrapping is necessary.
     * If the object is an array or `Collection`, returns an equivalent `JSONArray`.
     * If the object is a `Map`, returns an equivalent `JSONObject`.
     * If the object is a primitive wrapper type or `String`, returns the object.
     * Otherwise if the object is from a `java` package, returns the result of `toString`.
     * If wrapping fails, returns null.
     */
    fun wrap(o: Any?): Any?
    {
        if (o == null)
        {
            return JSONObject.NULL
        }

        if (o is JSONArray || o is JSONObject)
        {
            return o
        }

        if (o == JSONObject.NULL)
        {
            return o
        }

        try
        {
            if (o is Collection<*>)
            {
                val jsonArray = JSONArray()
                for (entry in (o as Collection<*>?)!!)
                {
                    jsonArray.put(wrap(entry))
                }
                return jsonArray
            }

            if (o.javaClass.isArray)
            {
                val length = java.lang.reflect.Array.getLength(o)
                val jsonArray = JSONArray()
                for (i in 0 until length)
                {
                    jsonArray.put(wrap(java.lang.reflect.Array.get(o, i)))
                }

                return jsonArray
            }

            if (o is Map<*, *>)
            {
                val `object` = JSONObject()

                for ((key1, value) in o)
                {
                    val key = key1 as? String ?: continue

                    try
                    {
                        `object`.put(key, wrap(value))
                    }
                    catch (e: JSONException)
                    {
                        e.printStackTrace()
                    }

                }

                return `object`
            }

            if (o is Boolean ||
                    o is Byte ||
                    o is Char ||
                    o is Double ||
                    o is Float ||
                    o is Int ||
                    o is Long ||
                    o is Short ||
                    o is String)
            {
                return o
            }

            if (o.javaClass.getPackage().name.startsWith("java."))
            {
                return o.toString()
            }

        }
        catch (ignored: Exception)
        {
        }

        return null
    }
}